generic.json

```diff
@@ -0,0 +1,33 @@
+{
+    "$schema": "http://json-schema.org/draft/2019-09/schema#",
+    "$id": "http://pop2020.dedis.ch/generic.json",
+    "type": "object",
+    "properties" : {
+        "jsonrpc": {
+            "const": "2.0"
+        },
+        "method": {
+            "description": "Operation to be performed",
+            "type": "string",
+            "enum": ["subscribe","unsubscribe","publish","message","catchup"]
+        },
+        "params": {
+            "description": "Content of the request",
+            "type": "object"
+        },
+        "result": {
+            "description": "For responses, present on success, absent on error",
+            "type": "integer"
+        },
+        "error": {
+            "description": "For responses, absent on success, present on error",
+            "type": "object"
+        },
+        "id": {
+            "description": "Unique identifier for the request",
+            "type": "integer"
+        }
```

Comment:
you're combining request and response fields.

An object which has both method and error fields will "comply" with this schema, which is undesirable.

You might want to "cheat" and have a look at this:
https://github.com/fge/sample-json-schemas/tree/master/jsonrpc2.0

Reply:
Ok I think I understand now! The idea for the "top-level schema" is to match ANY valid client request (or server answer in a different file) not matter if it is a "createLao message" or a "propagate a message message" for example?

I think we were both thinking of JsonSchema the wrong way (bottom to top instead of top to bottom)

Reply:
I didn't find (looked hard enough...) how to $ref between different files, both your early files and the link you pasted here only use $ref within a single file, which is not really useful to avoid code duplication, so I skipped it.

Now that I'm up to date with $ref between files, I'll update the files.

Also I like the fact of having a generic structure which can catch both request and response. But fully agree that we don't want to catch it a hybrid of both, but that could be fixed with the oneOf/not, etc. rather than making two arborescences.

Reply:
You don't need two arborescences. I just finished writing the schema for a server answer. Everything fits in a single file (since all types are very simple). Take a look at it:

```json
{
	"$schema": ???,
	"$id": ???,
	"title": "Match a JsonRpc 2.0 server answer",
	"description": "Match a positive or negative server answer",
	"type": "object",
	"properties": {
		"jsonrpc": { "description": "[String] JsonRpc version", "const": "2.0", "$comment": "should always be \"2.0\"" },
		"result": { 
			"description": "In case of positive answer, result of the client query",
			"anyOf": [
				{ "type": "integer", "minimum": 0, "maximum": 0 },
				{ "type": "array", "items": { "type": "string" } }
			],
			"$comment": "Note: this field is absent if the answer is negative"
		},
		"error": {
			"description": "In case of negative answer, error generated by the client query",
			"type": "object",
			"properties": {
				"code": { "type": "integer", "minimum": -5, "maximum": -1 },
				"description": { "type": "string" }
			},
			"additionalProperties": false,
			"required": ["code", "description"],
			"$comment": "Note: this field is absent if the answer is positive"
		},
		"id": { "type": "integer" }
	},
	"additionalProperties": false,
	"allOf": [
		{ "required": ["jsonrpc", "id"] },
		{ 
			"oneOf": [
				{ "required": ["result"] },
				{ "required": ["error"] }
			] 
		}
	],
	"$comment": "Note: required fields: (\"jsonrpc\" \\and (\"result\" \\or \"error\") \\and \"id\")"
}
```

If you prefer, I can still put both schemas in one file, but it might now be necessary. What do you think?

Reply:
Don't worry updating the `$ref`, I will do it when defining the other empty files

Reply:
Ok, it's not two "arborescence" because the return part is short, but it's still two different structures with no common parent. But your code works, but it's a matter of preference. I like having the directory structure matching the json structure for consistency and ease of reaching the file needed, so that each inner layer (object in json) is in another subfolder.

What I was going for was this at the top-layer: 
```
{
    "$schema": "http://json-schema.org/draft/2019-09/schema#",
    "$id": "http://pop2020.dedis.ch/generic.json",
    "type": "object",
    "properties" : {
        "jsonrpc": {
            "const": "2.0"
        },
        "oneOf": [
            {
                "anyOf": [
                    { "$ref": "method/subscribe.json#/properties" },
                    { "$ref": "method/unsubscribe.json#/properties" },
                    { "$ref": "method/publish.json#/properties" },
                    { "$ref": "method/broadcast.json#/properties" },
                    { "$ref": "method/catchup.json#/properties" }
                ]
            },
            {
                "anyOf": [
                    { "$ref": "returns/return.json#/properties" },
                    { "$ref": "returns/returnCatchup.json#/properties" }
                ]
            }
        ],
        "id": {
            "description": "Unique identifier for the request",
            "type": "integer"
        }

    },
    "required": ["jsonrpc", "oneOf"]
}
```

and this in the return layer:
```
{
    "$comment": "all requests except for broadcast and return expect a return as an ack",
    "$schema": "http://json-schema.org/draft/2019-09/schema#",
    "$id": "http://pop2020.dedis.ch/returns/return.json",
    "type": "object",
    "properties" : {
        "oneOf": [
            {
                "result": {
                    "description": "For responses, present on success, absent on error",
                    "type": "integer"
                }
            },
            {
                "error": {
                    "description": "For responses, absent on success, present on error",
                    "type": "object",
                    "$ref": "errors/error.json#/properties"
                }
            }
        ]

    },
    "required": ["oneOf"]
}
```

maybe @pierluca can chime in and state which is cleaner

Reply:
I think a directory structure is better, since our protocol is still subject to change and it is pretty complex already.